addons/netfox/rollback/rollback-synchronizer.gd

```diff
@@ -258,44 +327,65 @@ func _submit_input(input: Dictionary, tick: int):
 
 				var old_input = _inputs.get(t, {}).get(property)
 				var new_input = sanitized[property][i]
-				
+
 				if old_input == null:
 					# We received an array of current and previous inputs, merge them into our history.
 					_inputs[t] = _inputs.get(t, {})
 					_inputs[t][property] = new_input
-					_earliest_input = min(_earliest_input, t)
+					_earliest_input_tick = min(_earliest_input_tick, t)
 	else:
 		_logger.warning("Received invalid input from %s for tick %s for %s" % [sender, tick, root.name])
 
 @rpc("any_peer", "unreliable_ordered", "call_remote")
-func _submit_state(state: Dictionary, tick: int):
-	if tick > NetworkTime.tick:
-		# This used to be weird, but is now expected due to estimating remote time
-		# push_warning("Received state from the future %s / %s - adding nonetheless" % [tick, NetworkTime.tick])
-		pass
-	
-	if tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state >= 0:
+func _submit_full_state(state: Dictionary, tick: int):
+	if tick < NetworkTime.tick - NetworkRollback.history_limit:
 		# State too old!
-		_logger.error("Received state for %s, rejecting because older than %s frames" % [tick, NetworkRollback.history_limit])
+		_logger.error("Received full state for %s, rejecting because older than %s frames" % [tick, NetworkRollback.history_limit])
 		return
 
 	var sender = multiplayer.get_remote_sender_id()
-	var sanitized = {}
-	for property in state:
-		var property_entry = _property_cache.get_entry(property)
-		var value = state[property]
-		var state_owner = property_entry.node.get_multiplayer_authority()
-		
-		if state_owner != sender:
-			_logger.warning("Received state for node owned by %s from %s, sender has no authority!" \
-				% [state_owner, sender])
-			continue
-		
-		sanitized[property] = value
+	var sanitized = _sanitize_by_authority(state, sender)
 	
-	if sanitized.size() > 0:
-		_states[tick] = PropertySnapshot.merge(_states.get(tick, {}), sanitized)
-		# _latest_state = max(_latest_state, tick)
-		_latest_state = tick
-	else:
+	if sanitized.size() == 0:
+		# State is completely invalid
 		_logger.warning("Received invalid state from %s for tick %s" % [sender, tick])
```

Comment:
I suggest renaming the warning message to "Received completely empty state (should be full state) from %s for tick %s"

Reply:
The state is empty because it has no valid properties, i.e. the state itself is invalid.

Reply:
yup and this information should be exposed to the console instead of "received invalid state"

Reply:
No. The log message says that the state is invalid, because the state is invalid. The symptom ( state is empty after sanitization ) is not the same as the cause ( received state is invalid ). The line discussed is logging the cause, not the symptom. I am not changing that line to log the symptom.